packages/components/modules/profiles/web/ProfileComponent/ReportButtonWithDialog/index.tsx

```diff
@@ -0,0 +1,231 @@
+import { useReportCreateMutation } from '../../../common/graphql/mutations/ReportCreate'
+import { ReportTypeListQuery } from '../../../common/graphql/queries/ReportTypeList'
+import { TypeButton } from './styled'
+import { ReportButtonWithDialogProps } from './types'
+
+type AllReportTypes = NonNullable<ReportTypeListQueryType['response']['allReportTypes']>
+type ReportType = NonNullable<AllReportTypes['edges'][number]>
+type ReportTypeNode = NonNullable<ReportType['node']>
+type ReportTypeSubType = NonNullable<ReportTypeNode['subTypes']['edges'][number]>
+type ReportTypeSubTypeNode = ReportTypeSubType['node']
+
+const ReportButtonWithDialog: FC<ReportButtonWithDialogProps> = ({ targetId, handleClose }) => {
+  const { allReportTypes } = useLazyLoadQuery<ReportTypeListQueryType>(ReportTypeListQuery, {
+    topLevelOnly: true,
+    targetObjectId: targetId,
+  })
+  const [isReportModalOpen, setIsReportModalOpen] = useState(false)
+  const [currentStep, setCurrentStep] = useState('report')
+  const [reportType, setReportType] = useState<ReportTypeNode>()
+  const [reportSubType, setReportSubType] = useState<ReportTypeSubTypeNode>()
+  const [reportText, setReportText] = useState('')
+
+  const [commitMutation, isMutationInFlight] = useReportCreateMutation()
+  const reportTypes = useMemo(
+    () => allReportTypes?.edges?.filter((edge) => edge?.node).map((edge) => edge?.node) || [],
+    [allReportTypes?.edges],
+  ).sort((a, b) => {
+    if (!a) return -1
+    if (!b) return 1
+    return a.key > b.key ? 1 : -1
+  })
+  const subTypes = useMemo(
+    () => reportType?.subTypes?.edges?.filter((edge) => edge?.node).map((edge) => edge?.node) || [],
+    [reportType?.subTypes?.edges],
+  )
+
+  const onClose = () => {
+    handleClose()
+    setCurrentStep('report')
+  }
+
+  const handleReport = () => {
+    if (isMutationInFlight || !targetId || !reportType) {
+      return
+    }
+    commitMutation({
+      variables: {
+        input: {
+          reportSubject: reportText,
+          reportTypeId: reportType.id,
+          targetObjectId: targetId,
+        },
+      },
+      onCompleted: (_, errors) => {
+        if (!errors) {
+          setCurrentStep('confirmation')
+          return
+        }
+        onClose()
+      },
+      onError: () => {
+        onClose()
+      },
+    })
+  }
+
+  const steps = [
+    {
+      name: 'report',
+      content: (
+        <>
+          <Box>
+            <Typography variant="h5">Why are you reporting this?</Typography>
+          </Box>
+          <Box>
+            <Typography variant="body2">
+              Your report is anonymous. If someone is in immediate danger, call the local emergency
+              services - don&apos;t wait.
+            </Typography>
+          </Box>
+          <Divider />
+          <Box display="flex" flexDirection="column">
+            {reportTypes?.map((type) => (
+              <TypeButton
+                key={type?.id}
+                onClick={() => {
```

Comment:
extract a method for this instead of writing here directly

Reply:
Done